chainio/txmgr/geometric/geometric.go

```diff
@@ -0,0 +1,609 @@
+	}
+	if params.GasTipMultiplier == 0 {
+		params.GasTipMultiplier = defaultParams.GasTipMultiplier
+	}
+}
+
+func NewGeometricTxnManager(
+	ethClient ethBackend,
+	wallet wallet.Wallet,
+	logger logging.Logger,
+	metrics Metrics,
+	params GeometricTxnManagerParams,
+) *GeometricTxManager {
+	fillParamsWithDefaultValues(&params)
+	return &GeometricTxManager{
+		ethClient: ethClient,
+		wallet:    wallet,
+		logger:    logger.With("component", "GeometricTxManager"),
+		metrics:   metrics,
+		params:    params,
+	}
+}
+
+// GetNoSendTxOpts This generates a noSend TransactOpts so that we can use
+// this to generate the transaction without actually sending it
+func (m *GeometricTxManager) GetNoSendTxOpts() (*bind.TransactOpts, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	from, err := m.wallet.SenderAddress(ctxWithTimeout)
+	if err != nil {
+		return nil, utils.WrapError("failed to get sender address", err)
+	}
+	return &bind.TransactOpts{
+		From:   from,
+		NoSend: true,
+		Signer: txmgr.NoopSigner,
+	}, nil
+}
+
+func newTxnRequest(tx *types.Transaction) *txnRequest {
+	return &txnRequest{
+		tx:          tx,
+		requestedAt: time.Now(),
+		txAttempts:  make([]*transaction, 0),
+	}
+}
+
+// Send is used to sign and send a transaction to an evm chain.
+// It does gas estimation and gas price bumping to ensure the transaction gets mined,
+// but it does not do nonce management, so the tx argument must have the correct nonce already set.
+//
+// Send is blocking and safe to call concurrently, so sending multiple txs in parallel is safe.
+func (t *GeometricTxManager) Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+	return t.processTransaction(ctx, newTxnRequest(tx))
+}
+
+// processTransaction sends the transaction and runs the monitoring loop which will bump the gasPrice until the tx get
+// included.
+func (t *GeometricTxManager) processTransaction(ctx context.Context, req *txnRequest) (*types.Receipt, error) {
+	t.logger.Debug("new transaction",
+		"nonce", req.tx.Nonce(), "gasFeeCap", req.tx.GasFeeCap(), "gasTipCap", req.tx.GasTipCap(),
+	)
+	t.metrics.IncrementProcessingTxCount()
+	defer t.metrics.DecrementProcessingTxCount()
+
+	from, err := t.wallet.SenderAddress(ctx)
+	if err != nil {
+		return nil, utils.WrapError("failed to get sender address", err)
+	}
+
+	var txn *types.Transaction
+	var txID wallet.TxID
+	retryFromFailure := 0
+	for retryFromFailure < t.params.MaxSendTransactionRetry {
+		gasTipCap, err := t.estimateGasTipCap(ctx)
+		if err != nil {
+			return nil, utils.WrapError("failed to estimate gas tip cap", err)
+		}
+		txn, err = t.updateGasTipCap(ctx, req.tx, gasTipCap, from)
+		if err != nil {
+			return nil, utils.WrapError("failed to update gas price", err)
+		}
+		txID, err = t.wallet.SendTransaction(ctx, txn)
+		// fireblocks wallet uses go's net.Http client which returns a url.Error on timeouts
```

Comment:
geth also uses http client

Reply:
Oh snap you're right... https://github.com/ethereum/go-ethereum/blob/83e70aa3d00bbee9713b31768a30a7741eee8945/rpc/http.go#L231
Huh why is error handling so ugly lol? I remember reading https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully and thinking I had it figured out by using error interfaces everywhere, but clearly I haven't haha.

So hmmm... shouldn't every package just wrap the underlying package's errors and redefine their own? Basically abstracting errors so that their clients don't need to deal with the error dependencies of 5 package dependencies down?

Reply:
Also I'm confused when is context.DeadlineExceeded returned then? If they all use http.Do, wouldn't they all return url.Error with a Timeout() method instead of context.DeadlineExceeded?

Reply:
changed the comment to [0219820](https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/02198207e7af8b60f7abf044c5657aeffa74278c)
Let me know if this is better (at least hopefully not wrong).

Reply:
> If they all use http.Do, wouldn't they all return url.Error with a Timeout() method instead of context.DeadlineExceeded?

We don't know if they all return `url.Error` 😅